src/contracts/yas_pool.cairo

```diff
@@ -84,25 +185,304 @@ mod YASPool {
+                    TickBitmap::TickBitmapImpl::next_initialized_tick_within_one_word(
+                    @state_tick_bitmap, state.tick, self.tick_spacing.read(), zero_for_one
+                );
+
+                // ensure that we do not overshoot the min/max tick, as the tick bitmap is not aware of these bounds
+                if step_tick_next < TickMath::MIN_TICK() {
+                    step_tick_next = TickMath::MIN_TICK();
+                } else if step_tick_next > TickMath::MAX_TICK() {
+                    step_tick_next = TickMath::MAX_TICK();
+                };
+
+                // get the price for the next tick
+                let step_sqrt_price_next_X96 = TickMath::get_sqrt_ratio_at_tick(step_tick_next);
+
+                // compute values to swap to the target tick, price limit, or point where input/output amount is exhausted
+                let (ret_sqrt_price_X96, step_amount_in, step_amount_out, mut step_fee_amount) =
+                    SwapMath::compute_swap_step(
+                    state.sqrt_price_X96,
+                    if (zero_for_one && step_sqrt_price_next_X96 < sqrt_price_limit_X96)
+                        || (!zero_for_one && step_sqrt_price_next_X96 > sqrt_price_limit_X96) {
+                        sqrt_price_limit_X96
+                    } else {
+                        step_sqrt_price_next_X96
+                    },
+                    state.liquidity,
+                    state.amount_specified_remaining,
+                    self.fee.read()
+                );
+
+                state.sqrt_price_X96 = ret_sqrt_price_X96;
+
+                if exact_input {
+                    state
+                        .amount_specified_remaining -=
+                            IntegerTrait::<i256>::new(step_amount_in + step_fee_amount, false);
+                    state.amount_calculated = state.amount_calculated
+                        - IntegerTrait::<i256>::new(step_amount_out, false);
+                } else {
+                    state
+                        .amount_specified_remaining +=
+                            IntegerTrait::<i256>::new(step_amount_out, false);
+                    state.amount_calculated = state.amount_calculated
+                        + IntegerTrait::<i256>::new(step_amount_in + step_fee_amount, false);
+                };
+
+                // if the protocol fee is on, calculate how much is owed, decrement feeAmount, and increment protocolFee
+                if cache.fee_protocol > 0 {
+                    let delta: u256 = step_fee_amount / cache.fee_protocol.into();
+                    step_fee_amount -= delta;
+                    state.protocol_fee += delta.try_into().unwrap();
+                };
+
+                // update global fee tracker
+                if state.liquidity > 0 {
+                    state
+                        .fee_growth_global_X128 +=
+                            FullMath::mul_div(step_fee_amount, Q128, step_fee_amount);
```

Comment:
There might be an error here, specifically in the third parameter.
```
// update global fee tracker
if (state.liquidity > 0)
    state.feeGrowthGlobalX128 += FullMath.mulDiv(step.feeAmount, FixedPoint128.Q128, state.liquidity);
```
```suggestion
                            FullMath::mul_div(step_fee_amount, Q128, state.liquidity);
```

Reply:
You're right, good catch!

Reply:
✅